File: leetcode/0047.permutation2/typescript/001.backtracking.ts
function permuteUnique(nums: number[]): number[][] {
    nums.sort()
    const res: number[][] = [];
    const used = new Array(nums.length).fill(false);

    const backtrack = (path: number[]) => {
        if(path.length === nums.length) {
            res.push([...path]);
            return
        }

        let lastRemoved = -1000000;
        for(let i = 0; i < nums.length; i++) {
            if(used[i] || nums[i] === lastRemoved) {
                continue
            }

            used[i] = true;
            path.push(nums[i]);
            backtrack(path);

            used[i] = false;
            lastRemoved = path.pop();
        }
    }

    backtrack([]);

    return res;
};

File: leetcode/0047.permutation2/go/001.backtracking.go
func permuteUnique(nums []int) [][]int {
    result := [][]int{}
    used := make([]bool, len(nums))
    sort.Ints(nums)
    
    var backtrack func(path []int)
    backtrack = func(path []int) {
        if len(path) == len(nums) {
      
            result = append(result, append([]int{}, path...))
            return
        }
        
        lastRemoved := -10000
        for i := 0; i < len(nums); i++ {
            if used[i] || (nums[i] == lastRemoved) {
                continue
            }
            
            used[i] = true
            path = append(path, nums[i])

            backtrack(path)
            lastRemoved = path[len(path)-1]
            path = path[:len(path)-1]
            used[i] = false
        }
    }
    
    backtrack([]int{})
    return result
}
